Codex: starting post-sync-fix

CI failure class: **code_bug** (code bug — must fix)
```
test	Run tests	2026-04-16T12:45:51.6791359Z ERROR tests/cycle/test_freeze_performance.py::test_freezing_100_candidates_under_three_seconds - AssertionError: assert ['0001', '000...0004', '0005'] == ['0001', '000...0003', '0004']
test	Run tests	2026-04-16T12:45:51.6801451Z ERROR tests/cycle/test_freeze_performance.py::test_concurrent_candidate_insert_does_not_change_frozen_snapshot - AssertionError: assert ['0001', '000...0004', '0005'] == ['0001', '000...0003', '0004']
test	Run tests	2026-04-16T12:45:51.6807697Z ERROR tests/cycle/test_freeze_performance.py::test_freeze_ignores_pending_and_rejected_candidates - AssertionError: assert ['0001', '000...0004', '0005'] == ['0001', '000...0003', '0004']

```

--- a/src/data_platform/cycle/__init__.py
+++ b/src/data_platform/cycle/__init__.py
@@ -18,8 +18,10 @@ from data_platform.cycle.manifest import (
     ManifestAlreadyPublished,
     PublishManifestNotFound,
     get_latest_publish_manifest,
+    get_publish_manifest_for_snapshot,
     get_publish_manifest,
     publish_manifest,
+    validate_snapshot_id,
 )
 from data_platform.cycle.repository import (
     CycleAlreadyFrozen,
@@ -56,6 +58,8 @@ __all__ = [
     "get_cycle",
     "get_latest_publish_manifest",
     "get_publish_manifest",
+    "get_publish_manifest_for_snapshot",
     "publish_manifest",
     "transition_cycle_status",
+    "validate_snapshot_id",
 ]

--- a/src/data_platform/cycle/manifest.py
+++ b/src/data_platform/cycle/manifest.py
@@ -237,6 +237,50 @@ def get_latest_publish_manifest() -> CyclePublishManifest:
         engine.dispose()
 
 
+def get_publish_manifest_for_snapshot(
+    snapshot_id: int,
+    table_identifier: str,
+) -> CyclePublishManifest:
+    """Return the newest manifest that publishes table_identifier at snapshot_id."""
+
+    validated_snapshot_id = validate_snapshot_id(snapshot_id)
+    validated_table_identifier = _validate_formal_table_key(table_identifier)
+    engine = _create_engine()
+    try:
+        with engine.connect() as connection:
+            row = (
+                connection.execute(
+                    _text(
+                        f"""
+                    SELECT
+                        published_cycle_id,
+                        published_at,
+                        formal_table_snapshots
+                    FROM {CYCLE_PUBLISH_MANIFEST_TABLE}
+                    WHERE jsonb_extract_path_text(
+                        formal_table_snapshots,
+                        :table_identifier,
+                        'snapshot_id'
+                    ) = :snapshot_id
+                    ORDER BY published_at DESC, published_cycle_id DESC
+                    LIMIT 1
+                    """
+                    ),
+                    {
+                        "table_identifier": validated_table_identifier,
+                        "snapshot_id": str(validated_snapshot_id),
+                    },
+                )
+                .mappings()
+                .one_or_none()
+            )
+            if row is None:
+                raise PublishManifestNotFound()
+            return _manifest_from_row(row)
+    finally:
+        engine.dispose()
+
+
 def _normalize_snapshot_manifest(
     formal_table_snapshots: Mapping[str, object],
 ) -> dict[str, FormalTableSnapshot]:
@@ -295,7 +339,9 @@ def _validate_formal_table_key(table: object) -> str:
     return table
 
 
-def _validate_snapshot_id(snapshot_id: object) -> int:
+def validate_snapshot_id(snapshot_id: object) -> int:
+    """Validate a formal table snapshot id from API or manifest input."""
+
     if isinstance(snapshot_id, bool) or not isinstance(snapshot_id, int):
         msg = f"snapshot_id must be a positive integer: {snapshot_id!r}"
         raise InvalidFormalSnapshotManifest(msg)
@@ -303,6 +349,10 @@ def _validate_snapshot_id(snapshot_id: object) -> int:
         msg = f"snapshot_id must be a positive integer: {snapshot_id!r}"
         raise InvalidFormalSnapshotManifest(msg)
     return snapshot_id
+
+
+def _validate_snapshot_id(snapshot_id: object) -> int:
+    return validate_snapshot_id(snapshot_id)
 
 
 def _snapshot_payload(
@@ -347,7 +397,9 @@ __all__ = [
     "InvalidFormalSnapshotManifest",
     "ManifestAlreadyPublished",
     "PublishManifestNotFound",
+    "get_publish_manifest_for_snapshot",
     "get_latest_publish_manifest",
     "get_publish_manifest",
     "publish_manifest",
+    "validate_snapshot_id",
 ]

--- a/src/data_platform/serving/formal.py
+++ b/src/data_platform/serving/formal.py
@@ -8,15 +8,15 @@ from typing import Final
 import pyarrow as pa  # type: ignore[import-untyped]
 
 from data_platform.cycle.manifest import (
-    CYCLE_PUBLISH_MANIFEST_TABLE,
     CyclePublishManifest,
     FormalTableSnapshot,
+    InvalidFormalSnapshotManifest,
     PublishManifestNotFound,
-    _manifest_from_row,
     get_latest_publish_manifest,
+    get_publish_manifest_for_snapshot,
     get_publish_manifest,
+    validate_snapshot_id,
 )
-from data_platform.cycle.repository import _create_engine, _text
 from data_platform.serving.catalog import load_catalog
 from data_platform.serving.reader import _validate_identifier
 
@@ -102,11 +102,17 @@ def get_formal_by_snapshot(snapshot_id: int, object_type: str) -> FormalObject:
     """Read a formal object only if snapshot_id has been published in a manifest."""
 
     table_identifier = formal_table_identifier(object_type)
-    validated_snapshot_id = _validate_snapshot_id(snapshot_id)
-    manifest = _get_publish_manifest_for_snapshot(
-        validated_snapshot_id,
-        table_identifier,
-    )
+    validated_snapshot_id = _validate_formal_snapshot_id(snapshot_id)
+    try:
+        manifest = get_publish_manifest_for_snapshot(
+            validated_snapshot_id,
+            table_identifier,
+        )
+    except PublishManifestNotFound as exc:
+        raise FormalSnapshotNotPublished(
+            validated_snapshot_id,
+            table_identifier,
+        ) from exc
     snapshot = _snapshot_from_manifest(manifest, table_identifier)
     if snapshot.snapshot_id != validated_snapshot_id:
         raise FormalSnapshotNotPublished(validated_snapshot_id, table_identifier)
@@ -161,49 +167,11 @@ def _read_formal_snapshot(table_identifier: str, snapshot_id: int) -> pa.Table:
     raise TypeError(msg)
 
 
-def _get_publish_manifest_for_snapshot(
-    snapshot_id: int,
-    table_identifier: str,
-) -> CyclePublishManifest:
-    engine = _create_engine()
+def _validate_formal_snapshot_id(snapshot_id: int) -> int:
     try:
-        with engine.connect() as connection:
-            rows = (
-                connection.execute(
-                    _text(
-                        f"""
-                    SELECT
-                        published_cycle_id,
-                        published_at,
-                        formal_table_snapshots
-                    FROM {CYCLE_PUBLISH_MANIFEST_TABLE}
-                    ORDER BY published_at DESC, published_cycle_id DESC
-                    """
-                    )
-                )
-                .mappings()
-                .all()
-            )
-    finally:
-        engine.dispose()
-
-    for row in rows:
-        manifest = _manifest_from_row(row)
-        snapshot = manifest.formal_table_snapshots.get(table_identifier)
-        if snapshot is not None and snapshot.snapshot_id == snapshot_id:
-            return manifest
-
-    raise FormalSnapshotNotPublished(snapshot_id, table_identifier)
-
-
-def _validate_snapshot_id(snapshot_id: int) -> int:
-    if isinstance(snapshot_id, bool) or not isinstance(snapshot_id, int):
-        msg = f"snapshot_id must be a positive integer: {snapshot_id!r}"
-        raise ValueError(msg)
-    if snapshot_id < 1:
-        msg = f"snapshot_id must be a positive integer: {snapshot_id!r}"
-        raise ValueError(msg)
-    return snapshot_id
+        return validate_snapshot_id(snapshot_id)
+    except InvalidFormalSnapshotManifest as exc:
+        raise ValueError(str(exc)) from exc
 
 
 __all__ = [

--- a/tests/serving/test_formal.py
+++ b/tests/serving/test_formal.py
@@ -1,10 +1,13 @@
 from __future__ import annotations
 
+import os
+from collections.abc import Generator, Mapping
 from dataclasses import FrozenInstanceError, fields, is_dataclass
-from datetime import UTC, datetime
+from datetime import UTC, date, datetime
 import importlib.util
 from pathlib import Path
 from typing import Any
+from uuid import uuid4
 
 import pytest
 
@@ -168,7 +171,7 @@ def test_by_snapshot_reads_only_published_snapshot(
     monkeypatch.setattr(formal_module, "load_catalog", lambda: catalog)
     monkeypatch.setattr(
         formal_module,
-        "_get_publish_manifest_for_snapshot",
+        "get_publish_manifest_for_snapshot",
         published_lookup,
     )
 
@@ -208,12 +211,70 @@ def test_by_snapshot_rejects_unpublished_current_head_before_reading_table(
     monkeypatch.setattr(formal_module, "load_catalog", fail_load_catalog)
     monkeypatch.setattr(
         formal_module,
-        "_get_publish_manifest_for_snapshot",
+        "get_publish_manifest_for_snapshot",
         unpublished_lookup,
     )
 
     with pytest.raises(formal_module.FormalSnapshotNotPublished):
         formal_module.get_formal_by_snapshot(unpublished_snapshot_id, "recommendation_set")
+
+
+def test_by_snapshot_uses_db_backed_manifest_lookup(
+    formal_postgres_env: str,
+    monkeypatch: pytest.MonkeyPatch,
+    formal_module: Any,
+    pa_module: Any,
+) -> None:
+    target_snapshot_id = 98123
+    unpublished_snapshot_id = 98124
+    target_cycle_id = "CYCLE_20260201"
+
+    for day in range(1, 26):
+        _publish_formal_manifest(
+            date(2026, 1, day),
+            {"formal.other_object": 70000 + day},
+        )
+    _publish_formal_manifest(
+        date(2026, 2, 1),
+        {FORMAL_IDENTIFIER: target_snapshot_id},
+    )
+    for day in range(2, 10):
+        _publish_formal_manifest(
+            date(2026, 2, day),
+            {"formal.other_object": 80000 + day},
+        )
+
+    monkeypatch.setattr(
+        formal_module,
+        "load_catalog",
+        lambda: _fake_snapshot_catalog(
+            pa_module,
+            expected_snapshot_id=target_snapshot_id,
+        ),
+    )
+
+    formal_object = formal_module.get_formal_by_snapshot(
+        target_snapshot_id,
+        "recommendation_set",
+    )
+
+    assert formal_object.cycle_id == target_cycle_id
+    assert formal_object.snapshot_id == target_snapshot_id
+    assert formal_object.payload.column("version").to_pylist() == [
+        f"snapshot-{target_snapshot_id}"
+    ]
+    assert formal_object.payload.column("score").to_pylist() == [target_snapshot_id]
+
+    monkeypatch.setattr(
+        formal_module,
+        "load_catalog",
+        lambda: pytest.fail("unpublished snapshot must not read formal table head"),
+    )
+    with pytest.raises(formal_module.FormalSnapshotNotPublished):
+        formal_module.get_formal_by_snapshot(
+            unpublished_snapshot_id,
+            "recommendation_set",
+        )
 
 
 def test_missing_manifest_raises_formal_manifest_not_found(
@@ -314,3 +375,138 @@ def make_manifest(
             )
         },
     )
+
+
+@pytest.fixture()
+def postgres_dsn() -> Generator[str]:
+    admin_dsn = os.environ.get("DATABASE_URL") or os.environ.get("DP_PG_DSN")
+    if not admin_dsn:
+        pytest.skip("PostgreSQL formal serving tests require DATABASE_URL or DP_PG_DSN")
+
+    sqlalchemy = pytest.importorskip(
+        "sqlalchemy",
+        reason="PostgreSQL formal serving tests require SQLAlchemy",
+    )
+    runner_module = pytest.importorskip(
+        "data_platform.ddl.runner",
+        reason="PostgreSQL formal serving tests require the migration runner",
+    )
+    make_url = pytest.importorskip(
+        "sqlalchemy.engine",
+        reason="PostgreSQL formal serving tests require SQLAlchemy",
+    ).make_url
+    sqlalchemy_error = pytest.importorskip(
+        "sqlalchemy.exc",
+        reason="PostgreSQL formal serving tests require SQLAlchemy",
+    ).SQLAlchemyError
+
+    admin_engine = _create_engine(admin_dsn, isolation_level="AUTOCOMMIT")
+    database_name = f"dp_formal_serving_test_{uuid4().hex}"
+    try:
+        with admin_engine.connect() as connection:
+            connection.execute(sqlalchemy.text(f'CREATE DATABASE "{database_name}"'))
+    except sqlalchemy_error as exc:
+        admin_engine.dispose()
+        pytest.skip(
+            "PostgreSQL formal serving tests require permission to create "
+            f"test databases: {exc}"
+        )
+
+    test_dsn = str(
+        make_url(runner_module._sqlalchemy_postgres_uri(admin_dsn)).set(
+            database=database_name,
+        )
+    )
+    try:
+        yield test_dsn
+    finally:
+        with admin_engine.connect() as connection:
+            connection.execute(
+                sqlalchemy.text(
+                    """
+                    SELECT pg_terminate_backend(pid)
+                    FROM pg_stat_activity
+                    WHERE datname = :database_name
+                      AND pid <> pg_backend_pid()
+                    """
+                ),
+                {"database_name": database_name},
+            )
+            connection.execute(sqlalchemy.text(f'DROP DATABASE IF EXISTS "{database_name}"'))
+        admin_engine.dispose()
+
+
+@pytest.fixture()
+def migrated_postgres_dsn(postgres_dsn: str) -> str:
+    runner_module = pytest.importorskip(
+        "data_platform.ddl.runner",
+        reason="PostgreSQL formal serving tests require the migration runner",
+    )
+    runner_module.MigrationRunner().apply_pending(postgres_dsn)
+    assert runner_module.MigrationRunner().apply_pending(postgres_dsn) == []
+    return postgres_dsn
+
+
+@pytest.fixture()
+def formal_postgres_env(
+    migrated_postgres_dsn: str,
+    monkeypatch: pytest.MonkeyPatch,
+) -> Generator[str]:
+    monkeypatch.setenv("DP_PG_DSN", migrated_postgres_dsn)
+    yield migrated_postgres_dsn
+
+
+def _publish_formal_manifest(
+    cycle_date: date,
+    snapshots: Mapping[str, int],
+) -> None:
+    from data_platform.cycle import (
+        create_cycle,
+        publish_manifest,
+        transition_cycle_status,
+    )
+
+    cycle = create_cycle(cycle_date)
+    for status in ("phase0", "phase1", "phase2", "phase3"):
+        transition_cycle_status(cycle.cycle_id, status)
+    publish_manifest(cycle.cycle_id, snapshots)
+
+
+def _fake_snapshot_catalog(
+    pa_module: Any,
+    *,
+    expected_snapshot_id: int,
+) -> object:
+    class FakeScan:
+        def to_arrow(self) -> Any:
+            return pa_module.table(
+                {
+                    "version": [f"snapshot-{expected_snapshot_id}"],
+                    "score": [expected_snapshot_id],
+                },
+                schema=formal_schema(pa_module),
+            )
+
+    class FakeTable:
+        def scan(self, *, snapshot_id: int) -> FakeScan:
+            assert snapshot_id == expected_snapshot_id
+            return FakeScan()
+
+    class FakeCatalog:
+        def load_table(self, table_identifier: str) -> FakeTable:
+            assert table_identifier == FORMAL_IDENTIFIER
+            return FakeTable()
+
+    return FakeCatalog()
+
+
+def _create_engine(dsn: str, **kwargs: object) -> Any:
+    sqlalchemy = pytest.importorskip(
+        "sqlalchemy",
+        reason="PostgreSQL formal serving tests require SQLAlchemy",
+    )
+    runner_module = pytest.importorskip(
+        "data_platform.ddl.runner",
+        reason="PostgreSQL formal serving tests require the migration runner",
+    )
+    return sqlalchemy.create_engine(runner_module._sqlalchemy_postgres_uri(dsn), **kwargs)

